### ISSUE TITLE
New equal subscription instantly ready

### ISSUE THREAD

**Issue by radekmie** [meteor/meteor#7878]
## TL;DR;
```js
// Now:
Meteor.subscribe('subscription', arg1, arg2, ..., argN, () => {
    Meteor.subscribe('subscription', arg1, arg2, ..., argN).ready(); // false
});

// Suggested:
Meteor.subscribe('subscription', arg1, arg2, ..., argN, () => {
    Meteor.subscribe('subscription', arg1, arg2, ..., argN).ready(); // true
});
```

## Background
Probably most of React apps are using [`react-meteor-data`](https://github.com/meteor/react-packages/tree/devel/packages/react-meteor-data) or [`react-komposer`](https://github.com/kadirahq/react-komposer) to subscribe and fetch data. If you were working with one of them, you've probably seen some similar code:

```js
export default composer (props, onData) {
    const subscriptionA = Meteor.subscribe('a', props.a);
    const subscriptionB = Meteor.subscribe('b', props.b);
    // More subscriptions

    if (!subscriptionA.ready() || !subscriptionB.ready()) {
        // Loading state
    }

    // Ready state
}
```

Now let's assume, that we are subscribing the same subscriptions with the same parameters in other containers. Even if our _equal_ subscriptions are ready, we have to wait for a DDP `ready` message, so loading state is impossible to be skipped - loader can blink and disappear instantly.

## Workaround
I've implemented a rough workaround that wraps `Meteor.subscribe` with the following logic:

0. Subscribe normally.
0. Check if there's already an _equal_ subscription.
0. If there is at least one, mark handle as ready _(same as a DDP `ready` message)_.
0. Return the handle.

After applying it in few projects, some part of views is instantly loaded - loading screens are gone!

## Problem?
I've dived deeply into the [`ddp-client`](https://github.com/meteor/meteor/blob/devel/packages/ddp-client/) package and even if server rejects our subscription - ex. we can't subscribe after 12AM - being `ready` instantly [is not a problem because `nosub` message will clean subscription gracefully](https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js#L573-L576).

## Solution
I want to implement my workaround directly in `Meteor.subscribe`. It will be similar to this [reactive subscriptions mechanism](https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js#L573-L576) _(but without `sub.inactive`)_.

Thoughts?

**Comment by stubailo** [comment]
Hmm, so we already have some logic that does a similar thing inside a Tracker autorun: https://github.com/meteor/meteor/blob/06e583c9f0cd1777365111641d39fe4566e5ca87/packages/ddp-client/livedata_connection.js#L573

I wonder if we can merge those two situations into one case. Basically, multiplex subscriptions on the client.

**Comment by radekmie** [comment]
I think, that we shouldn't share subscription id, but something like this would work:

```js
var equal = (...); // first subscription with equal name and parameters
if (equal) {
  if (equal.inactive) {
    // resubscribed in tracker - share id
  } else {
    // new subscription - could be instant ready
  }
} else {
  // new subscription
}
```

If there are no objections, I could make a PR for it.

**Comment by stubailo** [comment]
That sounds pretty good to me.

**Comment by quintstoffers** [comment]
What about when one stops a subscription that is equal to another subscription? I think the expected behaviour would be to only stop the subscription when _all_ references to it are gone.

**Comment by tmeasday** [comment]
Sorry to come late to the party here but I think this isn't the right approach. 

Hiding the fact that the second identically subscription runs by making it appear insta-ready just obscures what's going on from the developer.

In your use-case I would suggest using something like [`subs-manager`](https://atmospherejs.com/meteorhacks/subs-manager) which correctly ensures that you only have one copy of each subscription with a familiar interface.

Alternatively, we could consider changing `ddp-client` to mirror that behaviour (to only ever have one copy of any identical subscription active). However this is a big change and I'm not sure I can justify it, given that workarounds like `subs-manager` exist.

**Comment by radekmie** [comment]
In some rare cases _(I've stumbled upon few of them)_, `subs-manager` could break things:

```js
Meteor.publish('x', () => {
    // I know it's weird, but something similar happened to me
    if (Math.random() > 0.5) {
        return Collection.find({x: new Date()});
    }

    throw new Meteor.Error('Nope');
});
```

Now, with `subs-manager` or changed `ddp-client` to have only one copy, there's no call to the server, so subscription succeeds. In my version, it's instantly ready, but if server _rejects_ _(like in this case)_, then it's cancelled as soon as response come back from the server.

**Comment by tmeasday** [comment]
But in this case the UI has no way to know that the subscription actually failed, as `ready` went `true` straight away. I think if your publication is not idempotent in the way you have described, then in fact you do generally want to know when the second subscription is ready.

Of course you are free to ignore the readiness if you want to carry on regardless, but this is complex behavior that perhaps your subscription manager could handle.

**Comment by radekmie** [comment]
There's `stopCallback` and deprecated `errorCallback` - [those will be fired, when server responds](https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js#L1561-L1579).
